src/main/java/gregtech/common/covers/filter/BaseFilterContainer.java

```diff
@@ -0,0 +1,170 @@
+package gregtech.common.covers.filter;
+
+import gregtech.api.util.IDirtyNotifiable;
+
+import net.minecraft.nbt.NBTTagCompound;
+import net.minecraft.network.PacketBuffer;
+import net.minecraft.util.math.MathHelper;
+import net.minecraftforge.common.util.INBTSerializable;
+import net.minecraftforge.items.ItemStackHandler;
+
+import com.cleanroommc.modularui.api.widget.IWidget;
+import com.cleanroommc.modularui.screen.ModularPanel;
+import com.cleanroommc.modularui.value.sync.GuiSyncManager;
+import org.jetbrains.annotations.NotNull;
+
+public abstract class BaseFilterContainer<R, T extends Filter<R>> implements INBTSerializable<NBTTagCompound> {
+
+    private final IDirtyNotifiable dirtyNotifiable;
+    private int maxTransferSize = 1;
+    private T currentFilter;
+    private Runnable onFilterInstanceChange;
```

Comment:
Should mark these fields as `@Nullable`, and all associated methods and parameters should receive appropriate annotations as well.